## Machine Learning(DataScience-PGC,InternshalaTrainings) Project

## Problem Statement:
 * Create a machine learning model which will help the company in determining the salary of newly
 hired employees using the given data.

#                    TechWorks Consulting Company

In [1]:
pwd

'C:\\Users\\aryan\\Desktop\\New for Jupyter'

## Load the data

In [2]:
#IMPORT THE LIBRARIES
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
df=pd.read_csv('C:/Users/aryan/Desktop/New for Jupyter/ML case Study.csv',header=0)

In [4]:
df.head()

College             City       Role  Previous CTC  \
0        SVNIT Surat          Asansol    Manager       55523.0   
1         NIT Bhopal            Ajmer  Executive       57081.0   
2       IEM, Kolkata  Rajpur Sonarpur  Executive       60347.0   
3  KIIT, Bhubaneswar            Ajmer  Executive       49010.0   
4                DTU         Durgapur  Executive       57879.0   

   Previous job change  Graduation Marks  EXP (Month)       CTC  
0                    3                66           19  71406.58  
1                    1                84           18  68005.87  
2                    2                52           28  76764.02  
3                    2                81           33  82092.39  
4                    4                74           32  73878.10

In [5]:
df.shape

(1589, 8)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1589 entries, 0 to 1588
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   College              1589 non-null   object 
 1   City                 1589 non-null   object 
 2   Role                 1589 non-null   object 
 3   Previous CTC         1589 non-null   float64
 4   Previous job change  1589 non-null   int64  
 5   Graduation Marks     1589 non-null   int64  
 6   EXP (Month)          1589 non-null   int64  
 7   CTC                  1589 non-null   float64
dtypes: float64(2), int64(3), object(3)
memory usage: 99.4+ KB


In [7]:
college=pd.read_csv('C:/Users/aryan/Desktop/New for Jupyter/Colleges.csv',header=0)

In [8]:
cities=pd.read_csv('C:/Users/aryan/Desktop/New for Jupyter/Cities.csv',header=0)

In [9]:
college.head()

Tier 1          Tier 2                                      Tier 3
0     IIT Bombay  IIIT Bangalore  Ramaiah Institute of Technology, Bengaluru
1      IIT Delhi      IIIT Delhi                      TIET/Thapar University
2  IIT Kharagpur          IGDTUW                         Manipal Main Campus
3     IIT Madras     NIT Calicut                                 VIT Vellore
4     IIT Kanpur   IIITM Gwalior                             SRM Main Campus

In [10]:
cities.head()

Metrio City non-metro cities
0      Mumbai         Dehradun
1       Delhi         Durgapur
2     Kolkata          Asansol
3     Chennai         Rourkela
4   Bangalore        Kozhikode

In [11]:
Tier1=college["Tier 1"].tolist()
Tier2=college["Tier 2"].tolist()
Tier3=college["Tier 3"].tolist()

In [12]:
Tier1

['IIT Bombay',
 'IIT Delhi',
 'IIT Kharagpur',
 'IIT Madras',
 'IIT Kanpur',
 'IIT Roorkee',
 'IIT Guwahati',
 'IIIT Hyderabad',
 'BITS Pilani (Pilani Campus)',
 'IIT Indore',
 'IIT Ropar',
 'IIT BHU (Varanasi)',
 'IIT ISM Dhanbad',
 'DTU',
 'NSUT Delhi (NSIT)',
 'NIT Tiruchipally (Trichy)',
 'NIT Warangal',
 'NIT Surathkal (Karnataka)',
 'Jadavpur University',
 'BITS Pilani (Hyderabad Campus)',
 'BITS Pilani (Goa Campus)',
 'IIIT Allahabad',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan]

In [13]:
Tier2

['IIIT Bangalore',
 'IIIT Delhi',
 'IGDTUW',
 'NIT Calicut',
 'IIITM Gwalior',
 'IIIT Lucknow',
 'MNNIT Allahabad',
 'Punjab Engineering College',
 'DAIICT',
 'MNIT Jaipur',
 'NIT Durgapur',
 'VNIT Nagpur',
 'LNMIIT',
 'BIT Mesra',
 'SVNIT Surat',
 'NIT Jalandhar',
 'NIT Jamshedpur',
 'NIT Kurukshetra',
 'NIT Patna',
 'NIT Raipur',
 'NIT Bhopal',
 'NIT Rourkela',
 'NIT Silchar',
 'NIT Sikkim',
 'IIIT Jabalpur',
 'Jalpaiguri Government Engineering College',
 'IIEST/BESU Shibpur',
 'R.V. College of Engineering']

In [14]:
Tier3

['Ramaiah Institute of Technology, Bengaluru',
 'TIET/Thapar University',
 'Manipal Main Campus',
 'VIT Vellore',
 'SRM Main Campus',
 'KIIT, Bhubaneswar',
 'Amity University',
 'BMS College, Bangalore',
 'PES University, Bangalore',
 'USIT, Delhi',
 'MAIT, Delhi',
 'MSIT, Delhi',
 'Dayanand College of Engineering, Bangalore',
 'Heritage Institute of Technology, Kolkata',
 'IEM, Kolkata',
 'Techno India, Kolkata',
 'Jaypee, Noida',
 'Amrita University, Coimbatore',
 'Cummins College of Engineering for Women, Pune',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan]

In [15]:
for item in df.College:
    if item in Tier1:
        df['College'].replace(item,1,inplace=True)
    elif item in Tier2:
        df['College'].replace(item,2,inplace=True)
    elif item in Tier3:
        df['College'].replace(item,3,inplace=True)

In [16]:
df.head()

College             City       Role  Previous CTC  Previous job change  \
0        2          Asansol    Manager       55523.0                    3   
1        2            Ajmer  Executive       57081.0                    1   
2        3  Rajpur Sonarpur  Executive       60347.0                    2   
3        3            Ajmer  Executive       49010.0                    2   
4        1         Durgapur  Executive       57879.0                    4   

   Graduation Marks  EXP (Month)       CTC  
0                66           19  71406.58  
1                84           18  68005.87  
2                52           28  76764.02  
3                81           33  82092.39  
4                74           32  73878.10

In [17]:
metro = cities['Metrio City'].tolist()
non_metro = cities['non-metro cities'].tolist()

In [18]:
for item in df.City:
    if item in metro:
        df['City'].replace(item,1,inplace=True)
    elif item in non_metro:
        df['City'].replace(item,0,inplace=True)

In [19]:
df.head()

College  City       Role  Previous CTC  Previous job change  \
0        2     0    Manager       55523.0                    3   
1        2     0  Executive       57081.0                    1   
2        3     0  Executive       60347.0                    2   
3        3     0  Executive       49010.0                    2   
4        1     0  Executive       57879.0                    4   

   Graduation Marks  EXP (Month)       CTC  
0                66           19  71406.58  
1                84           18  68005.87  
2                52           28  76764.02  
3                81           33  82092.39  
4                74           32  73878.10

In [20]:
# Converting Categorical column 'Role' into numerical
df = pd.get_dummies(df, drop_first=True,dtype=int)

In [21]:
df

College  City  Previous CTC  Previous job change  Graduation Marks  \
0           2     0       55523.0                    3                66   
1           2     0       57081.0                    1                84   
2           3     0       60347.0                    2                52   
3           3     0       49010.0                    2                81   
4           1     0       57879.0                    4                74   
...       ...   ...           ...                  ...               ...   
1584        1     1       61285.0                    3                44   
1585        1     0       63140.0                    1                44   
1586        1     1       44907.0                    1                60   
1587        1     0       52054.0                    2                67   
1588        2     1       44353.0                    3                75   

      EXP (Month)       CTC  Role_Manager  
0              19  71406.58             1  
1              18  68005.87             0  
2              28  76764.02             0  
3              33  82092.39             0  
4              32  73878.10             0  
...           ...       ...           ...  
1584           60  77283.59             0  
1585           24  68416.86             0  
1586           19  70186.19             0  
1587           29  73394.09             0  
1588           18  70056.31             1  

[1589 rows x 8 columns]

In [22]:
# Check missing values
df.isna().sum()

College                0
City                   0
Previous CTC           0
Previous job change    0
Graduation Marks       0
EXP (Month)            0
CTC                    0
Role_Manager           0
dtype: int64

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1589 entries, 0 to 1588
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   College              1589 non-null   int64  
 1   City                 1589 non-null   int64  
 2   Previous CTC         1589 non-null   float64
 3   Previous job change  1589 non-null   int64  
 4   Graduation Marks     1589 non-null   int64  
 5   EXP (Month)          1589 non-null   int64  
 6   CTC                  1589 non-null   float64
 7   Role_Manager         1589 non-null   int32  
dtypes: float64(2), int32(1), int64(5)
memory usage: 93.2 KB


In [24]:
df.describe()

College         City  Previous CTC  Previous job change  \
count  1589.000000  1589.000000   1589.000000          1589.000000   
mean      1.975456     0.514160  55518.453744             2.528634   
std       0.838330     0.499957   6655.218445             1.123918   
min       1.000000     0.000000  36990.000000             1.000000   
25%       1.000000     0.000000  50518.000000             2.000000   
50%       2.000000     1.000000  55291.000000             3.000000   
75%       3.000000     1.000000  60109.000000             4.000000   
max       3.000000     1.000000  77911.000000             4.000000   

       Graduation Marks  EXP (Month)            CTC  Role_Manager  
count       1589.000000  1589.000000    1589.000000   1589.000000  
mean          59.855255    39.044682   75353.278798      0.206419  
std           14.935139    14.108875   12587.288237      0.404862  
min           35.000000    18.000000   53020.320000      0.000000  
25%           46.000000    26.000000   66902.350000      0.000000  
50%           60.000000    39.000000   73028.670000      0.000000  
75%           73.000000    51.000000   80588.670000      0.000000  
max           85.000000    64.000000  123416.990000      1.000000

## Checking Outliers

#####  I think there is no Outliers present. But if ,not in high amount so removing outlier is not making so much impact so i am thinking create model with outlier

# Apply ML Model

In [25]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [26]:
# Split data into dependent and Independent Variable
X = df.loc[:, df.columns != 'CTC']
y = df['CTC']

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [28]:
y_test

1079     74059.06
405      84692.16
1492     75028.75
239      71001.53
610      62426.39
          ...    
1023     67435.46
700      62927.79
486      75143.25
672      60479.67
1303    105077.70
Name: CTC, Length: 318, dtype: float64

## ML Model at test size = 0.2

In [29]:
# Create a LinearRegression model
linear_reg = LinearRegression()

# Fit the model to the training data
linear_reg.fit(X_train, y_train)

# Make predictions on the test data
linear_reg_pred = linear_reg.predict(X_test)

In [30]:
# Calculate and print the R-squared (r2) score
print("r2_score:",r2_score(y_test, linear_reg_pred))

# Calculate and print the Mean Absolute Error (MAE)
print("Mean Absolute Error :", mean_absolute_error(y_test, linear_reg_pred))

# Calculate and print the Mean Squared Error (MSE)
print("Mean Squared Error:", mean_squared_error(y_test, linear_reg_pred))

r2_score: 0.5933517958385095
Mean Absolute Error : 7191.231067500029
Mean Squared Error: 77362774.9495653


In [31]:
print(linear_reg.coef_)

[-5.17509158e+02  4.82800374e+02  4.22242899e-01  8.74331790e+00
  1.94023127e+00  2.43718541e+02  1.86020838e+04]


In [32]:
print(linear_reg.intercept_)

39229.5661268356


In [33]:
# Import the Ridge regression model
ridge = Ridge()

# Fit the model to training data
ridge.fit(X_train, y_train)

# Make prediction on test data
ridge_predict = ridge.predict(X_test)

# Calculate and print the R-squared (r2) score
print("r2_score:",r2_score(y_test, ridge_predict))

# Calculate and print the Mean Absolute Error (MAE)
print(" Mean Absolute Error:", mean_absolute_error(y_test, ridge_predict))

# Calculate and print the Mean Squared Error (MSE)
print("Mean Squared Error:", mean_squared_error(y_test, ridge_predict))

r2_score: 0.5926580862926117
 Mean Absolute Error: 7198.215276305912
Mean Squared Error: 77494749.7005428


In [34]:
print("Coef:",ridge.coef_)
print("Intercept:",ridge.intercept_)

Coef: [-5.16821192e+02  4.76920014e+02  4.22312444e-01  7.76895529e+00
  1.98601422e+00  2.43648895e+02  1.85111317e+04]
Intercept: 39248.221855171156


In [35]:
# Create Lasso regression with default parameters
lasso = Lasso()

# Fit model with train data
lasso.fit(X_train, y_train)

# Make prediction on test data
lasso_pred = lasso.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, lasso_pred))

# Calculate and print the Mean Absolute Error (MAE) to measure prediction accuracy
print("Mean Absolute Error:", mean_absolute_error(y_test, lasso_pred))

# Calculate and print the Mean Squared Error (MSE) to assess prediction accuracy
print("Mean Squared Error:", mean_squared_error(y_test, lasso_pred))

r2_score: 0.5933030911807144
Mean Absolute Error: 7191.7094203244
Mean Squared Error: 77372040.76567228


In [36]:
# Print the coefficients of the linear regression model
print("Coef:",lasso.coef_)

# Print the intercept of the linear regression model
print("Intercept:",lasso.intercept_)

Coef: [-5.16164197e+02  4.78595253e+02  4.22241674e-01  8.05923215e+00
  1.93522068e+00  2.43703723e+02  1.85954658e+04]
Intercept: 39233.08896423159


# Decision Tree

In [37]:
# Import DecisionTreeRegressor
from sklearn.tree import DecisionTreeRegressor

In [38]:
# Create a DecisionTreeRegressor model
dtr = DecisionTreeRegressor()

# Train the model using the training data
dtr.fit(X_train, y_train)

# Train the model using the training data
dtr_pred = dtr.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, dtr_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("Mean Absolute Error:", mean_absolute_error(y_test, dtr_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("Mean Squared Error:", mean_squared_error(y_test, dtr_pred))

r2_score: 0.3430257370697032
Mean Absolute Error: 8121.177484276731
Mean Squared Error: 124986048.21220124


In [39]:
# Create Decision tree with max depth = 4
dtr_tuned = DecisionTreeRegressor(max_depth=4)

# Fit model with train data
dtr_tuned.fit(X_train, y_train)

# Make prediction on test data
dtr_tuned_pred = dtr_tuned.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, dtr_tuned_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("Mean Absolute Error:", mean_absolute_error(y_test, dtr_tuned_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("Mean Squared Error:", mean_squared_error(y_test, dtr_tuned_pred))

r2_score: 0.5974129112852862
Mean Absolute Error: 6780.081301726802
Mean Squared Error: 76590168.16773771


In [40]:
# Import Random Forest and Grid Search from sklearn
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [41]:
# Create Random forest regression on default parameters
rnd = RandomForestRegressor()

# Fit model on train data
rnd.fit(X_train, y_train)

# Make prediction on test data
rnd_pred = rnd.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rnd_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("Mean Absolute Error:", mean_absolute_error(y_test, rnd_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("Mean Squared Error:", mean_squared_error(y_test, rnd_pred))

r2_score: 0.6401201527108638
Mean Absolute Error: 6353.9981245283025
Mean Squared Error: 68465330.35138373


In [42]:
# Create Random Forest regression with tuned parameters
rnd_tuned = RandomForestRegressor(n_jobs=-1, max_features=5, min_samples_split=3)

# Fit model on train data
rnd_tuned.fit(X_train, y_train)

# Make prediction on test data
rnd_tuned_pred = rnd_tuned.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rnd_tuned_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("Mean Absolute Error:", mean_absolute_error(y_test, rnd_tuned_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("Mean Squared Error:", mean_squared_error(y_test, rnd_tuned_pred))

r2_score: 0.6339551907380183
Mean Absolute Error: 6406.408296949437
Mean Squared Error: 69638183.35011102


In [43]:
# Parameters
params_grid = {"max_features": [4,5,6,7,8,9,10,20,30,40],
              "min_samples_split": [2,3,5,10,20]}

In [44]:
grid_search = GridSearchCV(rnd_tuned, params_grid, n_jobs=-1, cv=5)

In [45]:
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=RandomForestRegressor(max_features=5,
                                             min_samples_split=3, n_jobs=-1),
             n_jobs=-1,
             param_grid={'max_features': [4, 5, 6, 7, 8, 9, 10, 20, 30, 40],
                         'min_samples_split': [2, 3, 5, 10, 20]})

In [46]:
grid_search.best_params_

{'max_features': 4, 'min_samples_split': 3}

## Check model with test size=0.1

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 42)

In [48]:
# create Linear regression model with test size = 0.1
linear_reg = LinearRegression()

# Fit model with train data
linear_reg.fit(X_train, y_train)

# Prediction using test data
linear_reg_pred = linear_reg.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, linear_reg_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, linear_reg_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, linear_reg_pred))

print()

# Print the coefficients of the linear regression model
print("Coef:",linear_reg.coef_)

# Print the intercept of the linear regression model
print("Intercept:",linear_reg.intercept_)

r2_score: 0.6363216852464182
MAE: 7264.595222184898
MSE: 75992712.22135596

Coef: [-4.42911803e+02  5.25765245e+02  4.08212376e-01  9.22873354e+01
  2.31517208e+00  2.51617523e+02  1.89309912e+04]
Intercept: 39185.220906907896


In [49]:
# Create Ridge regression with test size = 0.1
ridge = Ridge()

# Fit model with train data
ridge.fit(X_train, y_train)

# Prediction using test data
ridge_predict = ridge.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, ridge_predict))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, ridge_predict))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, ridge_predict))

print()

# Print the coefficients of the linear regression model
print("Coef:",ridge.coef_)

# Print the intercept of the linear regression model
print("Intercept:",ridge.intercept_)

r2_score: 0.6356079191742243
MAE: 7273.607496917735
MSE: 76141857.81931205

Coef: [-4.42864404e+02  5.20710813e+02  4.08256354e-01  9.19186306e+01
  2.35030487e+00  2.51556414e+02  1.88488104e+04]
Intercept: 39203.13097141677


In [50]:
# Create Lasso regression model with default parameters
lasso = Lasso()

# Fit model on train data
lasso.fit(X_train, y_train)

# Prediction on test data
lasso_pred = lasso.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, lasso_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, lasso_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, lasso_pred))

print()

# Print the coefficients of the linear regression model
print("Coef:",lasso.coef_)

# Print the intercept of the linear regression model
print("Intercept:",lasso.intercept_)

r2_score: 0.6362725365808041
MAE: 7265.304096193648
MSE: 76002982.12266862

Coef: [-4.41583747e+02  5.21553553e+02  4.08206824e-01  9.16362699e+01
  2.30982336e+00  2.51603269e+02  1.89244902e+04]
Intercept: 39188.904632871956


In [51]:
# Create Decision tree regression with default parameters
dtr = DecisionTreeRegressor()

# Fit model on train data
dtr.fit(X_train, y_train)

# Prediction on test data
dtr_pred = dtr.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, dtr_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, dtr_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, dtr_pred))

r2_score: 0.4685299830047136
MAE: 7118.00786163522
MSE: 111053770.37167484


In [52]:
# Create DecisionTreeRegressor model with tuned parameter
dtr_tuned = DecisionTreeRegressor(max_depth=4)

# Fit model on train data
dtr_tuned.fit(X_train, y_train)

# Prediction using test data
dtr_tuned_pred = dtr_tuned.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, dtr_tuned_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, dtr_tuned_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, dtr_tuned_pred))

r2_score: 0.6302592573081193
MAE: 7013.486928870299
MSE: 77259492.01819375


In [53]:
# Create Random forest regressor model on default parameters
rnd = RandomForestRegressor()

#Fit model on train data
rnd.fit(X_train, y_train)

# prediction using test data
rnd_pred = rnd.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rnd_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, rnd_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, rnd_pred))

r2_score: 0.6709880039518583
MAE: 6547.042679245285
MSE: 68748982.05025321


In [54]:
# Create Random forest regressor model with max features = 5 and min sample split = 3
rnd_tuned = RandomForestRegressor(n_jobs=-1, max_features=5, min_samples_split=3)

# Fit model using train data
rnd_tuned.fit(X_train, y_train)

# Prediction using test data
rnd_tuned_pred = rnd_tuned.predict(X_test)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rnd_tuned_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, rnd_tuned_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, rnd_tuned_pred))

r2_score: 0.6815225792026403
MAE: 6468.573905659378
MSE: 66547720.90013672


## Perform Feature Scaling

In [55]:
X = df.loc[:, df.columns != 'CTC']
y = df['CTC']

In [56]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [57]:
from sklearn.preprocessing import StandardScaler

In [58]:
scaler = StandardScaler()

In [59]:
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [60]:
df_X_train = pd.DataFrame(X_train_scaled, columns=X_train.columns)
df_X_test = pd.DataFrame(X_test_scaled, columns = X_test.columns)

In [61]:
np.round(df_X_train.describe(), 1)

College    City  Previous CTC  Previous job change  Graduation Marks  \
count   1271.0  1271.0        1271.0               1271.0            1271.0   
mean      -0.0     0.0           0.0                 -0.0              -0.0   
std        1.0     1.0           1.0                  1.0               1.0   
min       -1.2    -1.0          -2.8                 -1.4              -1.7   
25%       -1.2    -1.0          -0.8                 -0.5              -0.9   
50%        0.0     1.0          -0.0                  0.4               0.0   
75%        1.2     1.0           0.7                  1.3               0.9   
max        1.2     1.0           3.4                  1.3               1.7   

       EXP (Month)  Role_Manager  
count       1271.0        1271.0  
mean          -0.0          -0.0  
std            1.0           1.0  
min           -1.5          -0.5  
25%           -0.9          -0.5  
50%           -0.0          -0.5  
75%            0.9          -0.5  
max            1.7           2.0

## Check Model after feature scaling

In [62]:
# Create LinearRegression model with scaled data and test size 0.2
lr_scaled = LinearRegression()

# Fit model on train data
lr_scaled.fit(X_train_scaled, y_train)

# Make prediction using test data
lr_scaled_pred = lr_scaled.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, lr_scaled_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, lr_scaled_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, lr_scaled_pred))

print()

# Print the coefficients of the linear regression model
print("Coef:",lr_scaled.coef_)

# Print the intercept of the linear regression model
print("Intercept:",lr_scaled.intercept_)

r2_score: 0.5933517958385082
MAE: 7191.231067500048
MSE: 77362774.94956557

Coef: [-435.09743711  241.31880694 2802.97570998    9.81979783   29.0192253
 3467.75304336 7471.39306988]
Intercept: 75284.838245476


In [63]:
# Create Ridge regressor after feature scaling with test size 0.2
r_scaled = Ridge()

# Fit model using train data
r_scaled.fit(X_train_scaled, y_train)

# Make prediction on test data
r_scaled_pred = r_scaled.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, r_scaled_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, r_scaled_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, r_scaled_pred))

r2_score: 0.5932150226490305
MAE: 7192.60467728189
MSE: 77388795.36074264


In [64]:
# Create Lasso regressor model after feature scaling and test size 0.2
l_scaled = Lasso()

# Fit model on train data
l_scaled.fit(X_train_scaled, y_train)

# Make prediction using test data
l_scaled_pred = l_scaled.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, l_scaled_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, l_scaled_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, l_scaled_pred))

r2_score: 0.5933169519706638
MAE: 7191.560845403443
MSE: 77369403.8201196


In [65]:
# Create Decision Tree regression after feature scaling with test size of 0.2
dt_scaled = DecisionTreeRegressor()

# Fit model on train data
dt_scaled.fit(X_train_scaled, y_train)

# Make prediction using test data
dt_scaled_pred = dt_scaled.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, dt_scaled_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, dt_scaled_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, dt_scaled_pred))

r2_score: 0.336607029331234
MAE: 8244.991603773586
MSE: 126207175.0662154


In [66]:
# Create Random Forest regressor model on default parameter after feature scaling with test size 0.2
rf_scaled = RandomForestRegressor()

# Fit model on train data
rf_scaled.fit(X_train_scaled, y_train)

# Prediction using test data
rf_scaled_pred = rf_scaled.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rf_scaled_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, rf_scaled_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, rf_scaled_pred))

r2_score: 0.6365811294427497
MAE: 6380.841845597484
MSE: 69138611.72292435


In [67]:
# Create Random Forest model after feature scaling with parameter tunned
rf_scaled_tuned = RandomForestRegressor(max_features=5, min_samples_split=3, n_jobs=-1)

# Fit model on train data
rf_scaled_tuned.fit(X_train_scaled, y_train)

# Prediction on test data
rf_scaled_tuned_pred = rf_scaled_tuned.predict(X_test_scaled)

# Calculate and print the R-squared (r2) score to evaluate model performance
print("r2_score:",r2_score(y_test, rf_scaled_tuned_pred))

# Calculate and print the mean absolute error(MSE) score to evaluate model performance
print("MAE:", mean_absolute_error(y_test, rf_scaled_tuned_pred))

# Calculate and print the Mean Squared Error (MSE) to evaluate prediction errors
print("MSE:", mean_squared_error(y_test, rf_scaled_tuned_pred))

r2_score: 0.6415235554383398
MAE: 6388.084624695891
MSE: 68198340.04315595


# TASKS

# Q1 Your views about the problem statement?

## ANS-- My View
* Context and Company Background: TechWorks Consulting, a company specializing in IT talent recruitment, and highlights its unique approach to matching skilled IT professionals with job opportunities.

* Data Description: The Dataset conatins information about colleges, cities, roles, previous experience, and salary. This information will be used to train and test the predictive model.

* Regression Task: The primary objective is to perform a regression task, where the aim is to predict a continuous variable, specifically the salary of newly hired employees.

* Role of Statistics: The role of statistics is to build and check the accuracy of the model.

# Q2 What will be your approach to solving this task?

## ANS-- Creating a Salary Prediction Model: A Systematic Approach
- **Data Understanding:**
  - Begin by thoroughly understanding the provided dataset, including its structure, columns, and the meaning of each variable. Gain insights into the data's distribution, summary statistics, and potential outliers.
- **Data Preprocessing:**
  - Handle Missing Values: Identify and address missing data by imputation or removal, ensuring that data is complete.
  - Convert Categorical Data: Transform categorical variables (e.g., "College" and "City") into numerical format.
  - Feature Selection: Use statistical techniques such as Lasso, Ridge, or correlation analysis to select the most relevant features for salary prediction.

- **Model Selection:**
    - Choose different regression models  to build and evaluate the predictive models.
- **Model Training and Evaluation:**
    - Split the dataset into training and testing sets to train the models and assess their performance.
    - Use appropriate evaluation metrics like Mean Squared Error (MSE), R-squared, and Mean Absolute Error (MAE) to measure the model's accuracy.
    - Experiment with different hyperparameters for each model and use cross-validation to avoid overfitting.
- **Model Comparison:**
    - Compare the performance of different models and select the one with the best accuracy and generalization.
- **Further Improvement:**
    - Consider additional techniques for model improvement, such as feature engineering, hyperparameter tuning, and ensemble methods.

# Q3 What were the available ML model options you had to perform this task?

# ANS-- The available ML model options had to perform on this task

**1. Linear Regression:**
- Linear regression is a simple and interpretable model that assumes a linear relationship between the features and the target variable (salary). It's a good starting point and can provide baseline performance.

**2. Ridge Regression and Lasso Regression:**
- Ridge and Lasso regression are regularization techniques that can be used to handle multicollinearity and prevent overfitting. They are variants of linear regression that add regularization terms to the cost function.

**3. Decision Trees:**
- Decision tree-based models, like Random Forest and Gradient Boosting, are capable of capturing non-linear relationships in the data. They can handle both numerical and categorical features and automatically deal with feature importance.

# Q4 Which model’s performance is best and what could be the possible reason for that?

# ANS-- Random Forest Model is best
#### To compare the performance of different machine learning models, we'll focus on the R-squared (r2_score) metric, which measures the goodness of fit of the models. A higher R-squared indicates a better fit to the data. Additionally, we'll consider Mean Absolute Error (MAE) and Mean Squared Error (MSE) for reference.

# In scenario we will not take tuned models

**Scenario 1 (ML Model at test size = 0.2) :**
- Linear Regression: r2_score = 0.5933
- Ridge: r2_score = 0.5926
- Lasso: r2_score = 0.5933
- Decision Tree: r2_score = 0.3430
- Random Forest: r2_score = 0.6401

**Scenario 2 (Check Model at test size = 0.1) :**
- Linear Regression: r2_score = 0.6363
- Ridge: r2_score = 0.6356
- Lasso: r2_score = 0.6362
- Decision Tree: r2_score = 0.4685
- Random Forest: r2_score = 0.6709

**Scenario 3 (Check Model after feature scaling) :**
- Linear Regression: r2_score = 0.5933
- Ridge: r2_score = 0.5932
- Lasso: r2_score = 0.5933
- Decision Tree: r2_score = 0.3366
- Random Forest: r2_score = 0.6365

## If i will run code the values is change all time when i run the code 
## So approx values i written

# Q5 What steps can you take to improve this selected model’s performance even further?

# ANS-- To  improve this  selected model performance, I can take these steps:
- **Increase the Number of Trees (Estimators):**
    Random Forest's performance often benefits from increasing the number of decision trees (estimators) in the ensemble.
- **Tune Hyperparameters:** Perform a more thorough hyperparameter tuning by experimenting with different values for parameters like max_depth, min_samples_split, min_samples_leaf, and max_features. Grid Search or Randomized Search can help find the optimal combination of hyperparameters.
- **Feature Selection:** Consider removing or reducing the importance of less informative features to improve the model's efficiency and potentially its performance.
